### ISSUE TITLE
Large host offloading overheads sometimes seen with Qwen3-30B-A3B on 1x DGXH100

### ISSUE THREAD

**Issue by rao-ashish** [verl-project/verl#6367]
### System Info

```
----------Python Info----------
Version      : 3.12.13
Compiler     : GCC 11.4.0
Build        : ('main', 'Mar  4 2026 09:23:07')
Arch         : ('64bit', 'ELF')
------------Pip Info-----------
Version      : 26.0.1
Directory    : /usr/local/lib/python3.12/dist-packages/pip
vllm         : 0.17.0
sglang       : not found.
ray          : 2.54.0
torch        : 2.10.0+cu129
----------verl Info-----------
Version      : 0.8.0.dev
Directory    : /opt/workspace/RL/verl/verl
Commit Hash  : 460ccf3c2f19ae537f7de53243c89df69d746f05
----------Platform Info----------
Platform     : Linux-5.15.0-1055-aws-x86_64-with-glibc2.35
system       : Linux
node         : pool0-0038
release      : 5.15.0-1055-aws
version      : #60~20.04.1-Ubuntu SMP Thu Feb 22 15:49:52 UTC 2024
----------Environment----------
CUDA Runtime : 12.9
CUDA Compiler : Cuda compilation tools, release 12.9, V12.9.86
----------System Info----------
CPU Memory      : 1999.96 GB
GPU Count       : 8
GPU 1   Type    : NVIDIA H100 80GB HBM3
GPU 1   Memory  : 79.65 GB
GPU 2   Type    : NVIDIA H100 80GB HBM3
GPU 2   Memory  : 79.65 GB
GPU 3   Type    : NVIDIA H100 80GB HBM3
GPU 3   Memory  : 79.65 GB
GPU 4   Type    : NVIDIA H100 80GB HBM3
GPU 4   Memory  : 79.65 GB
GPU 5   Type    : NVIDIA H100 80GB HBM3
GPU 5   Memory  : 79.65 GB
GPU 6   Type    : NVIDIA H100 80GB HBM3
GPU 6   Memory  : 79.65 GB
GPU 7   Type    : NVIDIA H100 80GB HBM3
GPU 7   Memory  : 79.65 GB
GPU 8   Type    : NVIDIA H100 80GB HBM3
GPU 8   Memory  : 79.65 GB
```

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

- Job launch script: https://gist.github.com/rao-ashish/3ca410f36710075b17a9c83f2eff8770
- Verl commit: https://github.com/verl-project/verl/commit/460ccf3c2f19ae537f7de53243c89df69d746f05
- Container: verlai/verl:vllm017.latest
- vLLM version: 0.17.0
- Megatron version: 0.16.0

### Expected behavior

Large host offloading overheads are sometimes observed when post-training Qwen3-30B-A3B on 1x DGXH100 (job launch script / config attached above).

After running the script 5 times, I found that in 3 of the profiles, `update_actor` takes > 140s, with the majority of time being taken up by the host managing memory (calling into `cuMemCreate / cuMemMap / cuMemSetAccess / cuMemUnmap / cuMemRelease`) while the device does no work. Here is an example screenshot from such a profile:

<img width="1177" height="421" alt="Image" src="https://github.com/user-attachments/assets/d84d7e20-5851-4a57-b5b5-cd1d5b432009" />

In those 3 profiles, these host offloading overheads accounted for 85-90% of the `update_actor` stage, and 36%-46% of the total step time.

In the other 2 profiles, host offloading took a more reasonable amount of time (16-17s):

<img width="1014" height="448" alt="Image" src="https://github.com/user-attachments/assets/e77b649f-6bb3-4743-b1e5-a95a7f02e7dc" />

Raw data from the 5 profiles is collected below:
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-3a86da3a-7fff-d21d-8cf4-e2b7e7454d10"><div dir="ltr" style="margin-left:0pt;" align="center">
profile # | step time (s) | gen time (s) | gen step frac | old_log_prob time (s) | old_log_prob step frac | ref time (s) | ref step frac | update_weights time (s) | update_weights step frac | update_actor time (s) | update_actor step frac | update_actor host overhead time (s) | update_actor host overhead phase frac | update_actor host overhead step frac | real update time (s) | real update phase frac | update_actor offloading time (s) | update_actor offloading phase frac | update_actor offloading step frac | update_actor transfer time (s) | update_actor transfer phase frac | update_actor transfer step frac
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | 226.166 | 19.537 | 0.08638345286 | 74.468 | 0.329262577 | 67.368 | 0.2978697063 | 36.924 | 0.1632606139 | 28.864 | 0.1276230733 | 4.144 | 0.1435698448 | 0.01832282483 | 2.522 | 0.08737527716 | 16.817 | 0.5826288803 | 0.0743568883 | 12.673 | 0.4390590355 | 0.05603406348
2 | 204.743 | 18.325 | 0.08950244941 | 75.299 | 0.3677732572 | 49.728 | 0.2428800985 | 37.267 | 0.1820184329 | 24.12 | 0.1178062254 | 4.536 | 0.1880597015 | 0.02215460358 | 2.801 | 0.1161276949 | 17.152 | 0.7111111111 | 0.08377331582 | 12.616 | 0.5230514096 | 0.06161871224
3 | 357.784 | 19.701 | 0.0550639492 | 74.093 | 0.2070886345 | 74.234 | 0.207482727 | 8.051 | 0.02250240368 | 181.7 | 0.5078483107 | 162.705 | 0.8954595487 | 0.4547576191 | 3.41 | 0.01876719868 | 174.234 | 0.9589102917 | 0.4869809718 | 11.529 | 0.06345074298 | 0.03222335264
4 | 341.255 | 17.055 | 0.04997728971 | 76.48 | 0.2241139324 | 56.984 | 0.1669836339 | 8.294 | 0.0243044058 | 182.439 | 0.5346119471 | 160.437 | 0.879400786 | 0.4701381665 | 2.534 | 0.01388957405 | 174.922 | 0.958797187 | 0.512584431 | 14.485 | 0.07939640099 | 0.04244626452
5 | 341.435 | 19.28 | 0.05646755605 | 75.641 | 0.2215385066 | 70.175 | 0.205529603 | 32.137 | 0.09412333241 | 144.197 | 0.4223263579 | 123.334 | 0.8553159913 | 0.3612224874 | 2.526 | 0.01751770148 | 137.57 | 0.9540420397 | 0.4029170999 | 14.236 | 0.09872604839 | 0.04169461244

</div></b>

Column name key:

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-98120412-7fff-8f4c-4cc1-1b6e4f508213"><div dir="ltr" style="margin-left:0pt;" align="center">
Column name | Description
-- | --
profile # | The profile number for this set of experiments.
step time (s) | The time taken by the full RL step as measured from the TaskRunner.
gen time (s) | The time taken by the gen / rollout phase as measured from the TaskRunner.
gen step frac | The fraction of total step time taken up by the gen phase.
old_log_prob time (s) | The time taken for computing old_log_probs as measured from the TaskRunner.
old_log_prob step frac | The fraction of total step time taken up by old_log_probs.
ref time | The time taken to compute ref log probs for the KL loss as measured from the TaskRunner.
ref step frac | The fraction of total step time taken up by the ref phase.
update_weights time (s) | The time taken to update rollout weights as measured from the TaskRunner.
update_weights step frac | The fraction of total step time taken up by update_weights.
update_actor time (s) | The time taken to update the actor as measured from the TaskRunner.
update_actor step frac | The fraction of total step time taken up by update_actor.
update_actor host overhead time (s) | The time during update_actor when the device does no / very sparse work while the host also does nothing or only manages memory for host offloading (calls to cuMemCreate / cuMemMap / cuMemSetAccess / cuMemUnmap / cuMemRelease).
update_actor host overhead phase frac | The fraction of update_actor taken up by the host overhead described in the preceding row.
update_actor host overhead step frac | The fraction of total step time taken up by the host overhead described in the preceding 2 rows.
real update time (s) | The time during update_actor when the device does actual FLOPs. Measured as the time starting with the first embedding lookup through the full optimizer step.
real update phase frac | The fraction of update_actor when the device does actual FLOPs.
update_actor offloading time (s) | The time during update_actor when the host manages memory for host offloading and/or the device is only doing data transfers (no FLOPs).
update_actor offloading phase frac | The fraction of update_actor taken up by host overhead and D2H/H2D transfers as described in the preceding row.
update_actor offloading step frac | The fraction of total step time taken up by host overhead and D2H/H2D transfers during update_actor as described in the preceding 2 rows.
update_actor transfer time (s) | The time during update_actor when the device is actually busy doing data transfers. Computed as (update_actor offloading time) - (update_actor host overhead time).
update_actor transfer phase frac | The fraction of update_actor taken up by D2H/H2D transfers due to host offloading.
update_actor transfer step frac | The fraction of total step time taken up by D2H/H2D transfers during update_actor due to host offloading.

</div></b>

**Comment by Kirrito-k423** [comment]
Thank you for the detailed profiling data and analysis.

**Initial observations**:

The inconsistent host offloading overhead (4s vs 160s in your profiles) suggests this is related to PyTorch's lazy memory allocator behavior rather than a code bug. The  calls you see are part of the CUDA memory pool initialization, which happens lazily on first access.

**Probable causes**:

1. **CUDA memory pool fragmentation**: When host offloading is active, PyTorch may need to expand its memory pool. If the pool was initialized differently in different runs (due to varying workload patterns), subsequent runs may need to grow the pool more aggressively.

2. **Non-stateful weight offloading**: The  +  strategy loads parameters from CPU to GPU on each micro-batch. If the allocation pattern varies, CUDA's memory allocator may need to create new mappings.

3. **cuMemSetAccess overhead**: The  call you see (which clears GPU memory protection before CPU→GPU transfer) can be slow when the virtual address space is fragmented.

**Recommended diagnostics**:

1. Add  to your launch script to enable more aggressive memory reuse.

2. Try pre-warming the CUDA memory allocator by running a few dummy training steps before the profiling window.

3. Monitor  across runs to see if the peak memory footprint varies.

Would you be able to share the job launch script from your gist? The exact offloading configuration would help pinpoint whether the issue is in the optimizer state management or parameter loading path.
